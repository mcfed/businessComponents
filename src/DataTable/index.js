import React, {Component} from 'react';
import PropTypes from 'prop-types';
import {Table, Icon, Checkbox, Button, Row, Col, Form} from 'antd';

export class TableMenu extends Component {
  state = {
    visible: true,
    columns: []
  };
  //请求远程数据接口
  componentWillMount() {
    let {actions} = this.props;
  }
  // //处理表格提交后动作
  handleOk() {
    const {columns} = this.state;
    const {onSelectChange, onClosePopup} = this.props;
    //  console.log(columns)
    onSelectChange(columns);
    //  this.form.onSubmit()
    onClosePopup();
  }
  saveFormRef = form => (this.form = form);
  handleSubmit(values) {
    var {onSelectChange} = this.props;
    this.setState({
      columns: values
    });
  }
  handleChange(values) {
    const {onSelectChange} = this.props;
    this.setState({
      columns: values
    });
  }
  render() {
    const {
      form,
      initialValues,
      handleSubmit,
      children,
      defaultValue,
      columns,
      onClosePopup
    } = this.props;
    const saveFormRef = this.saveFormRef;
    const formFullItemLayout = {
      labelCol: {
        span: 6
      },
      wrapperCol: {
        span: 18
      }
    };
    return (
      <div
        className=''
        style={{
          width: 400,
          height: 200,
          padding: '10px',
          border: '1px solid #cfdae5',
          background: '#fff'
        }}>
        <Form onSubmit={handleSubmit} ref={saveFormRef} layout='inline'>
          <Checkbox.Group
            name='isShowArr'
            style={{width: '100%'}}
            defaultValue={defaultValue}
            onChange={this.handleChange.bind(this)}>
            <Row>
              {columns
                .filter(it => {
                  return it.title != '操作';
                })
                .map((it, idx) => {
                  return (
                    <Col span={8} key={idx}>
                      <Checkbox
                        value={it.key}
                        disabled={it.isRead == 1 ? true : false}>
                        {it.title}
                      </Checkbox>
                    </Col>
                  );
                })}
            </Row>
          </Checkbox.Group>
          <div style={{textAlign: 'right'}}>
            <Button size='small' onClick={onClosePopup}>
              取消
            </Button>
            <Button
              size='small'
              type='primary'
              onClick={this.handleOk.bind(this)}
              style={{marginLeft: '10px'}}>
              确定
            </Button>
          </div>
        </Form>
      </div>
    );
  }
}

class DataTable extends Component {
  state = {
    visible: false,
    columns: [],
    displayColumns: []
  };
  static defaultProps = {
    page: {},
    prefixCls: 'ant-table',
    pagination: {
      showTotal: total => `共 ${total} 条`,
      // showQuickJumper:true,
      size: 'middle',
      showSizeChanger: true,
      pageSizeOptions: ['10', '20', '50', '100']
    },
    //  scroll:{ y: 500 },
    style: {
      width: '100%'
    },
    showConfig: false,
    columns: []
  };
  showPopover() {
    this.setState({
      visible: true
    });
  }
  constructor(props) {
    super(props);
    this.state.columns = props.columns;
  }
  componentWillReceiveProps(nextProps) {
    let {columns} = nextProps;
    this.setState({
      columns: columns
    });
  }

  onSelectChange(checkedValues) {
    //console.log(checkedValues)
    this.setState({
      columns: this.state.columns.map(col => {
        if (checkedValues.indexOf(col.key) >= 0) {
          col.visible = true;
        } else {
          col.visible = false;
        }
        return col;
      })
    });
  }
  onClosePopup() {
    this.setState({
      visible: false
    });
  }
  onPopupVisibleChange(boolean) {
    // console.log('show',arguments)
    this.setState({
      visible: boolean
    });
  }

  renderTableMenu() {
    let {columns} = this.state;
    var defaultValue = columns
      .filter(
        col =>
          col.type != 'config' &&
          (col.visible === true || col.visible === undefined)
      )
      .map(col => col.key);
    return (
      <TableMenu
        defaultValue={defaultValue}
        columns={columns}
        onSelectChange={this.onSelectChange.bind(this)}
        onClosePopup={this.onClosePopup.bind(this)}
      />
    );
  }
  render() {
    let {pagination, showConfig, page, defaultSort, ...otherProps} = this.props;
    let {visible, columns} = this.state;
    let newColumns;
    // console.log(this.props,"datatablerender")
    if (showConfig) {
      // if(true){
      newColumns = columns.filter(col => {
        return col.visible == true || col.visible == undefined;
        // return true
      });
      // .concat([{
      //     title:" ",
      //     filterDropdown:(
      //       this.renderTableMenu()
      //     ),
      //     filterDropdownVisible:visible,
      //     onFilterDropdownVisibleChange:this.onPopupVisibleChange.bind(this),
      //     width:30,
      //     fixed:'right',
      //     type:'config'
      // }])
    } else {
      newColumns = columns;
    }
    /*增加是否有排序判断 增加列配置*/
    // console.log("defaultSort", defaultSort);
    if (defaultSort) {
      newColumns = newColumns.map(it => {
        defaultSort.columnKey == it.dataIndex
          ? (it = Object.assign(it, {defaultSortOrder: defaultSort.order}))
          : null;
        return it;
      });
    }
    // console.log("newColumns", newColumns);
    return (
      <Table
        key={defaultSort && defaultSort.columnKey}
        {...otherProps}
        columns={newColumns}
        pagination={!pagination ? false : Object.assign({}, pagination, page)}
      />
    );
  }
}

DataTable.propTypes = {
  /**
    表格列的配置描述 同antd table columns
  **/
  columns: PropTypes.array.isRequired,
  /**
  数据数组，同antd table dataSource
  **/
  dataSource: PropTypes.array.isRequired,
  /**
  默认排序参数  {columnKey,order} columnkey代表需要排序的columns的dataIndex order 选项为‘descend ascend’之一
  e.g. {columnKey:'name',order:'descend'}
  **/
  defaultSort: PropTypes.object,
  /**
  分页器  同antd table pagination
  **/
  pagination: PropTypes.oneOfType([PropTypes.bool, PropTypes.object])
};
export default DataTable;
