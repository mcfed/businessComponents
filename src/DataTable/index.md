#DataTable 组件

> 继承于 antd Table 实现并进行功能扩展

- 默认设定固定配置分页
- 扩展实现列配置功能实现隐藏显示列

## 默认配置

> 简化全局使用的配置参数

```
  page:{},
  prefixCls: 'ant-table',
  showQuickJumper:true,
  pagination:{
     showTotal:total => `共 ${total} 条`,
     // showQuickJumper:true,
     showSizeChanger:true,
     pageSizeOptions:['10','20','50','100'],
  },
  //  scroll:{ y: 500 },
  style:{
    width:"100%"
  },
  showConfig:false,
  columns:[]

```

## DataTable

> 功能扩展开启 showConfig 后，自动增加配置列 fixed 右侧

| 参数       |      说明      |  类型 | 默认值 |
| ---------- | :------------: | ----: | -----: |
| showConfig | 开启列配置模式 | Array |  false |

> showConfig 存在 BUG 还需要修订，表格基本上功能可正常使用

## TODO:

## REFACTOR:
