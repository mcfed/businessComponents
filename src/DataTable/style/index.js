import 'antd/es/table/style';
import 'antd/es/icon/style';
import 'antd/es/checkbox/style';
import 'antd/es/button/style';
import 'antd/es/row/style';
import 'antd/es/col/style';
import 'antd/es/form/style';
