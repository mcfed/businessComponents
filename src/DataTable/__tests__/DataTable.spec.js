import React from 'react';
import {shallow, mount} from 'enzyme';
import toJson from 'enzyme-to-json';
import {Table} from 'antd';
import DataTable, {TableMenu} from '../index';
import {config} from 'rxjs';

const setup = props => {
  // 通过 enzyme 提供的 shallow(浅渲染) 创建组件
  const wrapper = shallow(<DataTable {...props} />);
  return {
    props,
    wrapper
  };
};

const setupTableMenu = props => {
  // 通过 enzyme 提供的 shallow(浅渲染) 创建组件
  const wrapperTableMenu = shallow(<TableMenu {...props} />);
  return {
    props,
    wrapperTableMenu
  };
};

describe('快照测试', () => {
  it('DataTable 快照测试', () => {
    const {wrapper} = setup();
    expect(wrapper).toMatchSnapshot();
  });
});

describe('DataTable 组件是否渲染', () => {
  const {wrapper, props} = setup();
  // case1
  // 通过查找存在 Input,测试组件正常渲染
  it.skip('DataTable Component should be render', () => {
    //.find(selector) 是 Enzyme shallow Rendering 提供的语法, 用于查找节点
    // 详细用法见 Enzyme 文档 http://airbnb.io/enzyme/docs/api/shallow.html
    expect(wrapper.find('Table').exists()).toBe(true);
    // expect(toJson(wrapper)).toMatchSnapshot();
  });

  it.skip('DataTable Component should be pagination', () => {
    // expect(wrapper.find('Pagination').exists()).toBe(true);
  });
});

describe('state visible with some methods', () => {
  const {wrapper, props} = setup();

  it('default visible state is false', () => {
    expect(wrapper.state('visible')).toBe(false);
  });

  it('trggier showPopover visible to be true', () => {
    wrapper.instance().showPopover();
    expect(wrapper.state('visible')).toBe(true);
  });

  it('trggier onClosePopup visible to be false', () => {
    wrapper.instance().showPopover();
    wrapper.instance().onClosePopup();
    expect(wrapper.state('visible')).toBe(false);
  });

  it('trggier onPopupVisibleChange to any', () => {
    wrapper.instance().onPopupVisibleChange(true);
    expect(wrapper.state('visible')).toBe(true);
  });
});

describe('defaultSort ', () => {
  const columns = [
      {
        title: 'first',
        key: 1,
        dataIndex: 'first'
      },
      {
        title: 'second',
        key: 2,
        dataIndex: 'second'
      }
    ],
    defaultSort = {
      columnKey: 'first',
      order: 'descend'
    };
  const defaultProps = {
    rowKey: 'id',
    defaultSort: defaultSort,
    dataSource: [
      {
        first: 111,
        second: 222,
        id: 1
      },
      {
        first: 333,
        second: 444,
        id: 2
      }
    ],
    columns: columns
  };

  it('while defaultSort is exist', () => {
    const {wrapper, props} = setup({
      ...defaultProps
    });
    expect(wrapper.state('columns')).toEqual(
      columns.map(it => {
        defaultSort.columnKey == it.dataIndex
          ? (it = Object.assign(it, {defaultSortOrder: defaultSort.order}))
          : null;
        return it;
      })
    );
  });

  it('while defaultSort null', () => {
    const {wrapper, props} = setup({
      ...Object.assign(defaultProps, {defaultSort: {}})
    });
    expect(wrapper.state('columns')).toEqual(columns);
  });
});

describe('DataTable 方法测试全覆盖', () => {
  const columns = [
      {
        title: 'first',
        key: 1,
        dataIndex: 'first'
      },
      {
        title: 'second',
        key: 2,
        dataIndex: 'second',
        type: 'noConfig',
        visible: true
      }
    ],
    defaultSort = {
      columnKey: 'first',
      order: 'descend'
    },
    dataSource = [
      {
        first: 111,
        second: 222,
        id: 1
      },
      {
        first: 333,
        second: 444,
        id: 2
      }
    ];
  const defaultProps = {
    rowKey: 'id',
    defaultSort: defaultSort,
    dataSource: dataSource,
    columns: columns
  };

  it('showPopover 方法调用', () => {
    const {wrapper, props} = setup({
      ...defaultProps
    });
    wrapper.instance().showPopover();
    expect(wrapper.instance().state.visible).toEqual(true);
  });

  it('onClosePopup 方法调用', () => {
    const {wrapper, props} = setup({
      ...defaultProps
    });
    wrapper.instance().onClosePopup();
    expect(wrapper.instance().state.visible).toEqual(false);
  });

  it('onPopupVisibleChange 方法调用', () => {
    const {wrapper, props} = setup({
      ...defaultProps
    });
    wrapper.instance().onClosePopup(false);
    expect(wrapper.instance().state.visible).toEqual(false);
  });

  it('renderTableMenu 方法调用', () => {
    const {wrapper, props} = setup({
      ...defaultProps
    });
    wrapper.instance().state.columns = columns;
    const wrapperTable = wrapper.instance().renderTableMenu();
    expect(wrapperTable.props['columns']).toEqual(columns);
  });

  it('constructor 方法调用', () => {
    const {wrapper, props} = setup({
      ...defaultProps
    });
    wrapper.instance().constructor({columns: columns});
    expect(wrapper.instance().state.columns).toEqual(columns);
  });

  it('onSelectChange 方法调用', () => {
    const columnsExpect = [
      {
        title: 'first',
        key: 1,
        dataIndex: 'first',
        defaultSortOrder: 'descend',
        visible: true
      },
      {
        title: 'second',
        key: 2,
        dataIndex: 'second',
        type: 'noConfig',
        visible: false
      }
    ];
    const {wrapper, props} = setup({
      ...defaultProps
    });
    wrapper.instance().state.columns = columns;
    wrapper.instance().onSelectChange([1]);
    expect(wrapper.instance().state.columns).toEqual(columnsExpect);
  });
});

describe('TableMenu 组件是否渲染', () => {
  const columns = [
    {
      title: 'first',
      key: 1,
      dataIndex: 'first'
    },
    {
      title: 'second',
      key: 2,
      dataIndex: 'second',
      type: 'noConfig',
      visible: true
    }
  ];
  const defaultProps = {
    defaultValue: 1,
    columns: columns
  };
  const {wrapperTableMenu, props} = setupTableMenu({...defaultProps});
  it('wrapperTableMenu 是否render', () => {
    expect(wrapperTableMenu.find('div').exists()).toBe(true);
  });
});

describe('TableMenu 方法测试全覆盖', () => {
  const columns = [
    {
      title: 'first',
      key: 1,
      dataIndex: 'first'
    },
    {
      title: 'second',
      key: 2,
      dataIndex: 'second',
      type: 'noConfig',
      visible: true
    }
  ];
  const defaultProps = {
    defaultValue: 1,
    columns: columns,
    onSelectChange: jest.fn(),
    onClosePopup: jest.fn()
  };

  it('handleChange 方法调用', () => {
    const {wrapperTableMenu, props} = setupTableMenu({
      ...defaultProps
    });
    wrapperTableMenu.instance().state.columns = columns;
    wrapperTableMenu.instance().handleChange(columns);
    expect(wrapperTableMenu.instance().state.columns).toEqual(columns);
  });

  it('handleSubmit 方法调用', () => {
    const {wrapperTableMenu, props} = setupTableMenu({
      ...defaultProps
    });
    wrapperTableMenu.instance().state.columns = columns;
    wrapperTableMenu.instance().handleSubmit(columns);
    expect(wrapperTableMenu.instance().state.columns).toEqual(columns);
  });

  it('saveFormRef 方法调用', () => {
    const {wrapperTableMenu, props} = setupTableMenu({
      ...defaultProps
    });
    const form = {};
    wrapperTableMenu.instance().saveFormRef(form);
    expect(wrapperTableMenu.instance().form).toEqual(form);
  });

  it.skip('handleOk 方法调用', () => {
    const {wrapperTableMenu, props} = setupTableMenu({
      ...defaultProps
    });
    let onSelectChange = jest.fn();
    let onClosePopup = jest.fn();
    wrapperTableMenu.instance().state.onSelectChange = onSelectChange;
    wrapperTableMenu.instance().state.onClosePopup = onClosePopup;
    wrapperTableMenu.instance().handleOk();
    // expect(onSelectChange).toBeCalled();
    expect(onClosePopup).toBeCalled();
  });
});
