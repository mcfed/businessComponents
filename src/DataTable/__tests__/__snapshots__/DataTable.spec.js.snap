// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`快照测试 DataTable 快照测试 1`] = `
<withStore(Table)
  columns={Array []}
  pagination={
    Object {
      "pageSizeOptions": Array [
        "10",
        "20",
        "50",
        "100",
      ],
      "showSizeChanger": true,
      "showTotal": [Function],
      "size": "middle",
    }
  }
  prefixCls="ant-table"
  style={
    Object {
      "width": "100%",
    }
  }
/>
`;
